imports/lib/schemas/customTypes.ts

```diff
@@ -0,0 +1,91 @@
+// that there is no way to set a field only for a non-insert upsert operation.
+// This is a bit odd, but we're constrained by Mongo's lack of a $setOnNotInsert
+// operator, and this does match our historical behavior with SimplSchema.
+export const IsInsert = new Meteor.EnvironmentVariable<boolean>();
+export const IsUpdate = new Meteor.EnvironmentVariable<boolean>();
+export const IsUpsert = new Meteor.EnvironmentVariable<boolean>();
+
+// Allow overriding time for testing
+let clock: () => Date;
+export function setClock(newClock: () => Date) {
+  clock = newClock;
+}
+export function resetClock() {
+  clock = () => new Date();
+}
+resetClock();
+
+export const nonEmptyString = z.string().min(1);
+
+// In several of these auto-generated fields, we lie to zod about the output
+// type of the field - with type assertions, we say that they are non-nullable.
+// This will be true in the database, even though the transform functions don't
+// always return values (e.g. createdTimestamp will always be a Date, it'll just
+// get set on insertion and not on subsequent updates).
+
+export const stringId = z.string().regex(Id).optional()
+  .transform((v) => v as unknown as string);
+attachCustomJsonSchema(stringId, { bsonType: 'string', pattern: Id.source }, true);
+
+export const deleted = z.boolean().default(false);
+
+export const createdTimestamp = z.date()
+  .default(() => clock());
+attachCustomJsonSchema(createdTimestamp, { bsonType: 'date' }, true);
+
+export const updatedTimestamp = z.date().optional()
+  .transform((v) => {
+    if (v) return v;
+    if (IsUpdate.getOrNullIfOutsideFiber()) return clock();
+    return undefined;
+  });
+attachCustomJsonSchema(updatedTimestamp, { bsonType: 'date' });
+
+export const lastWriteTimestamp = z.date().optional()
+  .transform((v) => {
+    if (v) return v;
+    if (IsInsert.getOrNullIfOutsideFiber() ||
+      IsUpsert.getOrNullIfOutsideFiber() ||
+      IsUpdate.getOrNullIfOutsideFiber()) {
+      return clock();
+    }
+    return undefined as unknown as Date;
+  });
+attachCustomJsonSchema(lastWriteTimestamp, { bsonType: 'date' });
+
+export const createdUser = z.string().regex(Id).optional()
+  .transform((v) => {
+    if (v) return v;
+    try {
+      if (IsInsert.get() || IsUpsert.get()) return Meteor.userId()!;
+    } catch (e) {
+      /* ignore */
+    }
+    return undefined as unknown as string;
+  });
+attachCustomJsonSchema(createdUser, { bsonType: 'string', pattern: Id.source }, true);
+
+export const updatedUser = z.string().regex(Id).optional()
+  .transform((v) => {
+    if (v) return v;
+    try {
+      if (IsUpdate.get()) return Meteor.userId() ?? undefined;
+    } catch (e) {
+      /* ignore */
+    }
+    return undefined;
+  });
+attachCustomJsonSchema(updatedUser, { bsonType: 'string', pattern: Id.source });
+
+export const answer = z.string().transform((v) => answerify(v));
+attachCustomJsonSchema(answer, { bsonType: 'string', pattern: '^[^a-z]+$' });
```

Comment:
Doesn't have to happen in this PR, but I would prefer to move away from being prescriptive about what an answer can be.  I think the reality is that we have tried to be more strict about this than several writing teams, with the result being us having to on multiple occasions adapt on the fly out-of-band during hunt to try to express novel answer contents because we tried to enforce a rule that never really existed.

I feel like we'd get just as much value out of just uppercasing things by default in the UI, but letting the serialization/DB layer become more flexible here.  We could even allow operators to override the UI-imposed constraints.

We can keep the `answerify` transform for now for minimizing behavioral change in this PR, but I think the particular custom JSON schema on this line isn't needed.

Reply:
I'm fine with that, I'll make the change.

I have some vague ideas swirling around for how we should be handling answers. I agree that our prescriptiveness has gotten in our way in the last few years, but the flip side is that there's a reason for the prescriptiveness. Being able to require (e.g.) that answers match `/^[A-Z]+/` means that it's easier to do things like de-duplication and uniqueness checking, etc.

My not fully formed idea is that we should have a flag somewhere for "traditional" vs. "non-traditional" answers (terminology is a WIP), where traditional answers have to match `/^[A-Z]+/` (or at least be comparable if you upcase and strip non-alpha), and non-traditional answers are, like, literally byte strings or something. I'm still not sure entirely how to fit it together, but throwing it out in case it gets any ideas churning in your head.